From 5d0a0937be405d223a680b3cf48f2df375dd65c8 Mon Sep 17 00:00:00 2001
From: Marcel Reichmuth <marcel.reichmuth@netmodule.com>
Date: Fri, 4 Oct 2013 09:53:47 +0200
Subject: [PATCH 7/9] kernel: sdhci: add sd card support

Conflicts:
	arch/arm/boot/dts/zx3-pm3.dts
	arch/arm/configs/zx3_pm3_defconfig

Signed-off-by: Marcel Reichmuth <marcel.reichmuth@netmodule.com>
---
 drivers/mmc/host/sdhci.c |   34 ++++++++++++++++------------------
 1 file changed, 16 insertions(+), 18 deletions(-)

diff --git a/drivers/mmc/host/sdhci.c b/drivers/mmc/host/sdhci.c
index 6f0bfc0..8909695 100644
--- a/drivers/mmc/host/sdhci.c
+++ b/drivers/mmc/host/sdhci.c
@@ -973,9 +973,20 @@ static void sdhci_send_command(struct sdhci_host *host, struct mmc_command *cmd)
	int flags;
	u32 mask;
	unsigned long timeout;
+	u8 ctrl;

	WARN_ON(host->cmd);

+	/* Make sure that the card detect logic of the controller is disabled.
+		This is necessary, because there seems to be a bug in the hardware
+		which causes the controller to report a card removed event and
+		automatically disable power and clock of the card during a data
+		transfer.
+	*/
+	ctrl = sdhci_readb(host, SDHCI_HOST_CONTROL);
+	ctrl |= 0xc0;
+	sdhci_writeb(host, ctrl, SDHCI_HOST_CONTROL);
+
	/* Wait max 10 ms */
	timeout = 10;

@@ -1517,24 +1528,11 @@ static void sdhci_set_ios(struct mmc_host *mmc, struct mmc_ios *ios)

 static int sdhci_check_ro(struct sdhci_host *host)
 {
-	unsigned long flags;
-	int is_readonly;
-
-	spin_lock_irqsave(&host->lock, flags);
-
-	if (host->flags & SDHCI_DEVICE_DEAD)
-		is_readonly = 0;
-	else if (host->ops->get_ro)
-		is_readonly = host->ops->get_ro(host);
-	else
-		is_readonly = !(sdhci_readl(host, SDHCI_PRESENT_STATE)
-				& SDHCI_WRITE_PROTECT);
-
-	spin_unlock_irqrestore(&host->lock, flags);
-
-	/* This quirk needs to be replaced by a callback-function later */
-	return host->quirks & SDHCI_QUIRK_INVERTED_WRITE_PROTECT ?
-		!is_readonly : is_readonly;
+	/*
+	Write protect is not supported, so allways return false.
+	*/
+
+	return 0;
 }

 #define SAMPLE_COUNT	5
--
1.7.10.4
